File: Java/Problem2.java
import java.util.HashMap;
import java.util.Map;


class Node {

    Character key;
    Integer value;
    Node prev;
    Node next;

    Node() {
        this(null, null);
    }

    Node(Character key, Integer value) {
        this.key = key;
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}


class LRUCache {

    private int capacity;
    private Map <Character, Node> cache;
    private Node head;
    private Node tail;
    
    LRUCache(int capacity) {
        this.capacity = capacity;
        this.cache = new HashMap<>();
        this.head = new Node();
        this.tail = new Node();
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    void addNode(Node node) {
    node.prev = this.head;
    node.next = this.head.next;
    this.head.next.prev = node;
    this.head.next = node;
    }

    void removeNode(Node node) {
        Node prevNode = node.prev;
        Node nextNode = node.next;
        prevNode.next = nextNode;
        nextNode.prev = prevNode;
    }

    void moveToTop(Node node) {
        this.removeNode(node);
        this.addNode(node);
    }

    Node getAndRemoveTail() {
        Node node = this.tail.prev;
        this.removeNode(node);
        return node;
    }

    int get(int key) {
        Node node = this.cache.get((char) key);
        if (node == null) return -1;
        this.moveToTop(node);
        return node.value;
        
    }
    
    void put(int key, int value) {
        Node node = this.cache.get((char) key);

        if (node == null) {
            Node newNode = new Node((char) key, value);
            this.cache.put((char) key, newNode);
            this.addNode(newNode);
            if (this.cache.size() > this.capacity) {
                Node tail = this.getAndRemoveTail();
                this.cache.remove(tail.key);
            }
        } else {
            node.value = value;
            this.moveToTop(node);
        }

    }
}

File: Python/Problem2.py
class Node:

    def __init__(self, key=None, value=None):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:
    
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head

    def addNode(self, node):
        """Agrega un nodo justo después de head"""
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node

    def removeNode(self, node):
        """Desconecta un nodo de la lista"""
        prevNode = node.prev
        nextNode = node.next
        prevNode.next = nextNode
        nextNode.prev = prevNode

    def moveToTop(self, node):
        """Mueve un nodo a la posición más reciente (inmediatamente después de head)"""
        self.removeNode(node)
        self.addNode(node)

    def getAndRemoveTail(self):
        """Remueve y retorna el nodo menos recientemente usado (antes del tail)"""
        node = self.tail.prev
        self.removeNode(node)
        return node

    def get(self, key: int) -> int:
        node = self.cache.get(key)
        if not node:
            return -1
        self.moveToTop(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        node = self.cache.get(key)
        if not node:
            newNode = Node(key, value)
            self.cache[key] = newNode
            self.addNode(newNode)
            if len(self.cache) > self.capacity:
                tail = self.getAndRemoveTail()
                del self.cache[tail.key]
        else:
            node.value = value
            self.moveToTop(node)
